crates/matrix-sdk/src/send_queue/mod.rs

```diff
@@ -674,13 +722,171 @@ impl RoomSendQueue {
+            return 0;
+        };
+
+        let Some((cache_key, parent_is_thumbnail_upload)) = dependent_requests.iter().find_map(|r| {
+            if r.parent_transaction_id != txn_id {
+                return None;
+            }
+            as_variant!(&r.kind, DependentQueuedRequestKind::UploadFileOrThumbnail { cache_key, parent_is_thumbnail_upload, .. } => (cache_key.clone(), *parent_is_thumbnail_upload))
+        }) else {
+            return 0;
+        };
+
+        // If this is not a thumbnail upload, we're uploading a gallery and the
+        // dependent request is for the next gallery item.
+        if !parent_is_thumbnail_upload {
+            return 0;
+        }
+
+        if let Ok(cache) = room.client().event_cache_store().lock().await {
+            if let Ok(Some(content)) = cache.get_media_content(&cache_key).await {
+                content.len()
+            } else {
+                0
+            }
+        } else {
+            0
+        }
+    }
+
+    /// Try to creates an observable to watch a media's upload progress.
+    fn try_create_media_upload_progress_observable(
+        report_media_upload_progress: &Arc<AtomicBool>,
+        media_upload_info: &MediaUploadInfo,
+        related_txn_id: &Option<OwnedTransactionId>,
+        updates: &broadcast::Sender<RoomSendQueueUpdate>,
+    ) -> Option<SharedObservable<TransmissionProgress>> {
+        if !report_media_upload_progress.load(Ordering::SeqCst) {
+            return None;
+        }
+
+        if let Some(related_txn_id) = related_txn_id {
+            let progress: SharedObservable<TransmissionProgress> = Default::default();
+            let mut subscriber = progress.subscribe();
+
+            let report_media_upload_progress = Arc::clone(report_media_upload_progress);
+            let media_upload_info = media_upload_info.clone();
+            let related_to = related_txn_id.clone();
+            let updates = updates.clone();
+
+            // Watch and communicate the progress on a detached background task. Once
+            // the progress observable is dropped, next() will return None and the
+            // task will end.
+            spawn(async move {
+                while let Some(progress) = subscriber.next().await {
+                    if report_media_upload_progress.load(Ordering::SeqCst) {
+                        let _ = updates.send(RoomSendQueueUpdate::MediaUpload {
+                            related_to: related_to.clone(),
+                            file: None,
+                            index: media_upload_info.index,
+                            progress: compute_combined_progress(
+                                progress,
+                                media_upload_info.bytes,
+                                media_upload_info.uploaded,
+                                media_upload_info.pending,
+                            ),
+                        });
+                    }
+                }
+            });
```

Comment:
@bnjbvr [said](https://github.com/matrix-org/matrix-rust-sdk/pull/5008#discussion_r2151535506):

> Does it correctly terminate when the upload is cancelled? (needs a test!)

Reply:
According to the docs, `next` will return `None` once `progress` is dropped which should break the loop and end the task. I'm struggling a little to envision an automated test for this. I can only think of sending some synthetic `RoomSendQueueUpdate::MediaUploadTaskEnded` (or similar)?